File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        
        // i < 2^(n)
        vector<vector<int>> output; 

        for (int binval = 0; binval < (1 << nums.size()); binval++){
            vector<int> subset; 
            for (int index = 0; index < nums.size(); index++){
                if ((1 << index) & binval){
                    subset.push_back(nums[index]);
                } 
            }

            output.push_back(subset);
        }
        return output;
    }
};

File: 78-subsets/subsets.py


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:

        output = []
        curr = []

        def dfs(nums, curr, output, index):
            if index < 0: 
                return

            dfs(nums, curr, output, index - 1)

            # add current value, add to output, then go to n - 1
            curr.append(nums[index])
            output.append(curr[0::])
            dfs(nums, curr, output, index - 1)
            curr.pop()
        
        output.append([])
        dfs(nums, curr, output, len(nums) - 1)
        return output


